### PR TITLE
Teleport options

### PR DESCRIPTION
**Pull request by AdamPlenty**

### CODE REVIEW

**Comment by Loobinex** [comment]
Currently the behavior is inconsistent. I've got a testmap with a vampire trapped in a lair surrounded by dirt. So completely disconnected to the rest of the dungeon.
- When I press L, I teleport to the library
- When (from the initial position) I press F, I do not teleport to the fight, but to the dungeon heart.

Pressing F does work when there's a path to walk there.

File: src/front_input.c [hunk 1]
```diff
@@ -1180,57 +1181,83 @@ short get_creature_control_action_inputs(void)
...
+        if (is_game_key_pressed(Gkey_ZoomRoom00 + i, &val, false))
+        {
+            clear_key_pressed(val);
+            TeleDest = i + 1;
+        }
+        
+    }
+                if (is_key_pressed(KC_SEMICOLON,KMod_DONTCARE))
+                {
+                    TeleDest = 17;
```
Comment (Loobinex) [inline src/front_input.c:1247]: Adding a comment here to briefly say where it goes, would be nice.

**Comment by Loobinex** [comment]
You've skipped the last zoom key, 'zoom to annoyed creature', is that on purpose?

File: src/thing_creature.c [hunk 2]
```diff
@@ -1242,18 +1255,204 @@ void process_thing_spell_teleport_effects(struct Thing *thing, struct CastedSpel
...
+                switch(TeleDest)
+                {
+                    case 0:
+                    {
+                        desttng = thing_get(cctrl->lairtng_idx);
+                        break;
+                    }
+                    case 1:
+                    {
+                        rkind = RoK_TREASURE;
```
Comment (Loobinex) [inline src/thing_creature.c:1267]: You could save a lot of code here by removing all the cases with a room. So, case1, case2, etc.
Then have a
default:
{
    rkind=(zoom_key_room_order[TeleDest]);
}

This also makes this function more robust to other changes.

**Comment by AdamPlenty** [comment]
> 
> 
> Currently the behavior is inconsistent. I've got a testmap with a vampire trapped in a lair surrounded by dirt. So completely disconnected to the rest of the dungeon.
> 
>     * When I press L, I teleport to the library
> 
>     * When (from the initial position) I press F, I do not teleport to the fight, but to the dungeon heart.
> 
> 
> Pressing F does work when there's a path to walk there.

That's intentional. Do you think it'd be better if you could teleport to fights the creature can't otherwise get to?

File: src/thing_creature.c (hunk 2 above)
Comment (AdamPlenty) [reply src/thing_creature.c:1267]: > 
> 
> You've skipped the last zoom key, 'zoom to annoyed creature', is that on purpose?

You think that's worth including? I never thought that's a likely place someone would want to teleport a creature to.

**Comment by Loobinex** [comment]
Well, consistency is key. If you can teleport to rooms you can't otherwise get to, why not fights/CtA as well?

File: src/thing_creature.c (hunk 2 above)
Comment (Loobinex) [reply src/thing_creature.c:1267]: It's not really an important teleport location, but I think it would be nice for consistency sake. If you have 'press a zoom key in possession and teleport to go there', that's a lot nicer than 'press all but one of the zoom keys in possession to go there.'

File: src/thing_creature.c [hunk 3]
```diff
@@ -1242,18 +1255,204 @@ void process_thing_spell_teleport_effects(struct Thing *thing, struct CastedSpel
...
+                        }
+                        else
+                        {
+                            allowed = false;
+                        }
+                        break;
+                    }
+                    case 19:
+                    {
+                        desttng = find_hero_door_hero_can_navigate_to(thing);
```
Comment (Loobinex) [inline src/thing_creature.c:1413]: I would actually remove case 19, the hero door. It's too powerful an ability, as it allows you to actually discover the hero portals by teleporting to them all.

Comment (AdamPlenty) [reply src/thing_creature.c:1413]: It only works if the creature can get there normally. That's how find_hero_door_hero_can_navigate_to works.

Comment (Loobinex) [reply src/thing_creature.c:1413]: Well, 'being able to get there' is something else than 'having explored it'. For rooms this is not the case, for CtA, Annoyed Creature, Fight, you all have it explored.

Hero gates behind doors, across lava, or even simply in unexplored areas beyond heroes on guard posts, are not explored yet.

File: src/thing_creature.c (hunk 2 above)
Comment (AdamPlenty) [reply src/thing_creature.c:1267]: > 
> 
> You could save a lot of code here by removing all the cases with a room. So, case1, case2, etc.
> Then have a
> default:
> {
> rkind=(zoom_key_room_order[TeleDest]);
> }
> 
> This also makes this function more robust to other changes.

I tried that, but it appears to have broken this feature; for some reason I can't teleport to the Treasure Room any more. It does work for the Torture Chamber, but I haven't tested any other room.

Comment (Loobinex) [reply src/thing_creature.c:1267]: Send me the code that does not work, I'll have a look.
Did pressing T teleport you to the entrance by any chance?

**Comment by AdamPlenty** [comment]
> 
> 
> Well, consistency is key. If you can teleport to rooms you can't otherwise get to, why not fights/CtA as well?

The reason I made Call to Arms have the 'Can you get there otherwise?' requirement is because otherwise it can be abused; all you have to do is place a CtA right inside enemy territory, possess a creature, and cast Teleport and hey presto!

**Comment by Loobinex** [comment]
I hate how inconsistent teleport was in the first place,... normally you can cast CtA on a teleporting unit and get it to teleport to places anyway, but indeed when you have inaccessible spaces this is not the case. There's a [topic on this on keeperklan](https://keeperklan.com/threads/4729)

The current behavior is:
- Units can teleport only to places they can get to normally
- In possession, units can teleport to the heart/lair always. So you can teleport through locked doors no problem,...

The new behavior will introduce issues no matter what I see.
Think of this scenario (A):
1) You drop units into a fight outside your dungeon. The door reaching your dungeon is locked, so units cannot walk there.
2) You possess the vampire, fight until you are about to die, then teleport home.
3) When you can heal, you heal, so you can fight again. You now press F and teleport to the fight -> you cannot.

This is really confusing, that you can teleport only one way, and very hard to understand. Here you DO want to be able to teleport to all fights.

Now think of this scenario (B):
1) You have used a reveal special to see blue
2) When blue digs out a room, you drop an imp onto the dirt before blue claims it
3) When blue imps start to fight your red imp, you possess your vampire and teleport into the dungeon you cannot get to normally.

It is clear that in this scenario, you clearly should not be able to teleport into the dungeon. So you DO NOT want to be able to teleport to all fights.

I also think it's important that mistresses and vampires don't start teleporting to rooms you've put behind locked doors, since that would make dungeon management impossible.

How about these requirements:
- Outside of possession, the current behavior will stay, and vampires will teleport only to places they can access
- In possession you can teleport though all locked doors, but not to places that are truly inaccessible (like walled in enemy dungeons). This includes your own rooms.
- However, due to magical reasons, the exception to this is his lair, and the dungeon heart where the vampire can always teleport too.

**Comment by AdamPlenty** [comment]
> 
> 
> I hate how inconsistent teleport was in the first place,... normally you can cast CtA on a teleporting unit and get it to teleport to places anyway, but indeed when you have inaccessible spaces this is not the case. There's a [topic on this on keeperklan](https://keeperklan.com/threads/4729)
> 
> The current behavior is:
> 
>     * Units can teleport only to places they can get to normally
> 
>     * In possession, units can teleport to the heart/lair always. So you can teleport through locked doors no problem,...
> 
> 
> The new behavior will introduce issues no matter what I see.
> Think of this scenario (A):
> 
>     1. You drop units into a fight outside your dungeon. The door reaching your dungeon is locked, so units cannot walk there.
> 
>     2. You possess the vampire, fight until you are about to die, then teleport home.
> 
>     3. When you can heal, you heal, so you can fight again. You now press F and teleport to the fight -> you cannot.
> 
> 
> This is really confusing, that you can teleport only one way, and very hard to understand. Here you DO want to be able to teleport to all fights.
> 
> Now think of this scenario (B):
> 
>     1. You have used a reveal special to see blue
> 
>     2. When blue digs out a room, you drop an imp onto the dirt before blue claims it
> 
>     3. When blue imps start to fight your red imp, you possess your vampire and teleport into the dungeon you cannot get to normally.
> 
> 
> It is clear that in this scenario, you clearly should not be able to teleport into the dungeon. So you DO NOT want to be able to teleport to all fights.
> 
> I also think it's important that mistresses and vampires don't start teleporting to rooms you've put behind locked doors, since that would make dungeon management impossible.
> 
> How about these requirements:
> 
>     * Outside of possession, the current behavior will stay, and vampires will teleport only to places they can access
> 
>     * In possession you can teleport though all locked doors, but not to places that are truly inaccessible (like walled in enemy dungeons). This includes your own rooms.
> 
>     * However, due to magical reasons, the exception to this is his lair, and the dungeon heart where the vampire can always teleport too.

This should only affect teleporting in possession, not where creatures teleport of their own volition. How do you suggest we make doors an exception to the "blocking slab" rule? This sounds to me like we need another "can the creature get here?" function, but with doors excluded somehow.

**Comment by Loobinex** [comment]
According to mefisto(see the link), there's already such a function for when they try to take paydays.

**Comment by AdamPlenty** [comment]
> 
> 
> According to mefisto(see the link), there's already such a function for when they try to take paydays.

I've had a look, but I'm not sure where that is. I've noticed however, that creatures do head for a Treasure Room on paydays even if they can't get there, and only stop when they encounter a locked door.

**Comment by Loobinex** [comment]
I found the solution, use:
` if (creature_can_navigate_to(thing, &cta_pos, NavRtF_NoOwner)`

File: src/front_input.c [hunk 4]
```diff
@@ -1024,10 +1025,10 @@ short get_creature_control_action_inputs(void)
     struct Thing* thing = thing_get(player->controlled_thing_idx);
     if (menu_is_active(GMnu_CREATURE_QUERY1))
     {
-      if ( ( (is_key_pressed(KC_7,KMod_NONE) || (is_key_pressed(KC_NUMPAD7,KMod_NONE))) & (creature_instance_get_available_id_for_pos(thing,6) > 0) ) ||
```
Comment (Loobinex) [inline src/front_input.c:1027]: Are these changes still needed?

File: src/front_input.c [hunk 5]
```diff
@@ -1180,57 +1181,83 @@ short get_creature_control_action_inputs(void)
...
+        for (int i = 0; i <= 15; i++)
+        {
+            if (is_game_key_pressed(Gkey_ZoomRoom00 + i, &val, false))
+            {
+                clear_key_pressed(val);
+                TeleDest = i + 1;
+            }
+        
+        }
+        if (is_key_pressed(KC_SEMICOLON,KMod_DONTCARE))
```
Comment (Loobinex) [inline src/front_input.c:1245]: A comment at the end of these lines, what they do would be nice.

**Comment by Loobinex** [comment]
Seems pretty good at this point. Is there anything you feel like should still be done?
I'm still pretty convinced that teleporting to hero gates is an undesired feature.

**Comment by Loobinex** [comment]
Well, there's a count_player_rooms_of_type function, it tells you how much rooms of said type you have.
You could put this chunk of code inside a loop, and only if all room types cannot be accessed, go to the default:
```
                if (rkind > 0)
                {
                    room = nearest ? find_room_nearest_to_position(thing->owner, rkind, &thing->mappos, &distance) : room_get(find_next_room_of_type(thing->owner, rkind));
                }
                if (!room_is_invalid(room))
                {
                    room_pos.x.val = subtile_coord_center(room->central_stl_x);
                    room_pos.y.val = subtile_coord_center(room->central_stl_y);
                    allowed = creature_can_navigate_to(thing, &room_pos, NavRtF_NoOwner);
                    if (!allowed)
                    {
                        if (find_random_valid_position_for_thing_in_room(thing, room, &room_pos))
                        {
                            allowed = creature_can_navigate_to(thing, &room_pos, NavRtF_NoOwner);
                        }
                    }
                }
```

File: src/frontmenu_ingame_tabs.c [hunk 6]
```diff
@@ -1205,6 +1205,7 @@ void gui_go_to_next_creature_activity(struct GuiButton *gbtn)
 
 RoomIndex find_my_next_room_of_type(RoomKind rkind)
```
Comment (Loobinex) [inline src/frontmenu_ingame_tabs.c:1206]: I see you're modifying an original function here. Why, and isn't there stuff still using this?

File: src/thing_creature.c [hunk 7]
```diff
@@ -1241,19 +1255,162 @@ void process_thing_spell_teleport_effects(struct Thing *thing, struct CastedSpel
         if (thing_in_wall_at(thing, &pos))
         {
             const struct Coord3d* newpos = NULL;
+            struct Coord3d room_pos;
             {
```
Comment (Loobinex) [inline src/thing_creature.c:1259]: I know you didn't place it, but is this bracket doing anything here? It looks like a mistake to me.

File: src/thing_creature.c [hunk 8]
```diff
@@ -1241,19 +1255,162 @@ void process_thing_spell_teleport_effects(struct Thing *thing, struct CastedSpel
...
+                    }
+                    case 19:
+                    {
+                        desttng = find_hero_door_hero_can_navigate_to(thing);
+                        allowed = (!thing_is_invalid(desttng));
+                        break;
+                    }
+                    default:
+                    {
+                        rkind = zoom_key_room_order[TeleDest-1];
```
Comment (Loobinex) [inline src/thing_creature.c:1350]: I think it's wise to set some conditions here,... only do this if TeleDest-1 would return a room. Or check that there's no way this can go wrong.

File: src/front_input.c (hunk 4 above)
Comment (AdamPlenty) [reply src/front_input.c:1027]: I've no idea.

File: src/thing_creature.c [hunk 9]
```diff
@@ -1241,19 +1255,160 @@ void process_thing_spell_teleport_effects(struct Thing *thing, struct CastedSpel
...
+                default:
+                {
+                    rkind = zoom_key_room_order[TeleDest-1];
+                }
+            }
+            if (rkind > 0)
+            {
+                room = nearest ? find_room_nearest_to_position(thing->owner, rkind, &thing->mappos, &distance) : room_get(find_next_room_of_type(thing->owner, rkind));
+            }
+            if (!room_is_invalid(room))
```
Comment (Loobinex) [inline src/thing_creature.c:1356]: I think you need to nest this condition in the statement above. if rkind <= 0, there is no room.

File: src/frontmenu_ingame_tabs.c (hunk 6 above)
Comment (AdamPlenty) [reply src/frontmenu_ingame_tabs.c:1206]: I needed to be able to specify a player, so made a new function allowing me to do so, and made the original a wrapper.

Comment (Loobinex) [reply src/frontmenu_ingame_tabs.c:1206]: I see.

You can just remove the bit you commented out.

File: src/thing_creature.c [hunk 10]
```diff
@@ -1241,19 +1255,160 @@ void process_thing_spell_teleport_effects(struct Thing *thing, struct CastedSpel
...
+                }
+                case 19:
+                {
+                    desttng = find_hero_door_hero_can_navigate_to(thing);
+                    allowed = (!thing_is_invalid(desttng));
+                    break;
+                }
+                default:
+                {
+                    rkind = zoom_key_room_order[TeleDest-1];
```
Comment (Loobinex) [inline src/thing_creature.c:1349]: I thought I already made a comment here, but it is gone. I think you should look into making sure this only happens when teledest-1 can return a room. I've not looked into what values teledest can have, and what happens if it returns -1 or 18 or something.

File: src/thing_creature.c (hunk 8 above)
Comment (AdamPlenty) [reply src/thing_creature.c]: It shouldn't go wrong, as zoom_key_order is hard-coded. Do you think it should reset to 0 if the resulting value is not a room?
